Donation-platform/src/test/java/com/google/PersonalizedRecommendationsTest.java

```diff
@@ -11,269 +11,340 @@
 // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 // See the License for the specific language governing permissions and
 // limitations under the License.
-
+ 
 package com.google;
-
+ 
+import com.google.model.Charity;
+import com.google.model.Tag;
+import com.google.DbSetUpUtils;
 import java.util.Arrays;
 import java.util.Collection;
 import java.util.List;
+import java.util.ArrayList;
 import java.util.HashSet;
 import org.junit.Assert;
 import org.junit.Before;
+import org.junit.After;
 import org.junit.Test;
 import org.junit.runner.RunWith;
 import org.junit.runners.JUnit4;
-
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.tools.development.testing.LocalDatastoreServiceTestConfig;
+import com.google.appengine.tools.development.testing.LocalServiceTestHelper;
+ 
 /* Unit tests for PersonalizedRecommendations.java */
 @RunWith(JUnit4.class)
 public final class PersonalizedRecommendationsTest {
-
-  private static final Charity FA = HardCodedCharitiesAndTags.charities[0];
-  private static final Charity RC = HardCodedCharitiesAndTags.charities[1];
-  private static final Charity SJ = HardCodedCharitiesAndTags.charities[2];
-  private static final Charity NC = HardCodedCharitiesAndTags.charities[3];
-  private static final Charity YMCA = HardCodedCharitiesAndTags.charities[4];
-  private static final Charity ACLU = HardCodedCharitiesAndTags.charities[5];
-  private static final Charity AHA = HardCodedCharitiesAndTags.charities[6];
-
+ 
+  private Charity FA;
+  private Charity RC;
+  private Charity SJ;
+  private Charity NC;
+  private Charity YMCA;
+  private Charity ACLU;
+  private Charity AHA;
+ 
   private PersonalizedRecommendations recommendation;
-
+  private LocalServiceTestHelper helper;
+  private DatastoreService ds;
+  private DbCalls db;
+ 
   @Before
   public void setUp() {
-    // Reset each charity's trending score so that only tags are considered
-    // in testing the personalized ranking.
-    for (Charity charity : Arrays.asList(HardCodedCharitiesAndTags.charities)) {
-      charity.setTrendingScore(0);
+    helper = new LocalServiceTestHelper(new LocalDatastoreServiceTestConfig());
+    helper.setUp();
+    ds = DatastoreServiceFactory.getDatastoreService();
+    db = new DbCalls(ds);
+    DbSetUpUtils dbSetUp = new DbSetUpUtils(ds, db);
+    dbSetUp.populateDatabase();
+
+    try {
+      FA = db.getCharityByName("Feeding America");
+      RC = db.getCharityByName("Red Cross");
+      SJ = db.getCharityByName("St. Jude's");
+      NC = db.getCharityByName("Nature Conservancy");
+      YMCA = db.getCharityByName("YMCA");
+      ACLU = db.getCharityByName("ACLU");
```

Comment:
Add both charity names and objects in a list and iterate over them. 

for eg: 

charityNames = ["Feeding America", "Red Cross" etc] 
Iterate over charityNames and set trending score in same loop.

Reply:
@charu0291 I considered that, but the problem is I'm trying to save the charity objects with the instance variables FA, RC, SJ, etc. I tried storing these uninitialized charity objects in an array to iterate over but that causes nullpointer exceptions because they're not initialized. I could have two loops: one that adds all the objects once initialized and another one to set each charity's trending score to zero. I think my current method is more efficient in terms of time complexity, but using these two loops would help shorten my code. What do you think is the best approach here?

Reply:
for (Charity charityName: charities ) {
  charityObj = db.getCharityByName(charityName);
  charityObj.setTrendingScoreCharity(0);
}

Does above code give null pointerexception? Yes time complexity is same in both cases. Its more for readability and less lines of code.

Reply:
Yes, because there is no "charities" array/collection. I tried initializing one with the uninitialized objects FA, RC, SJ, etc, but that gave the nullpointer exception.

Reply:
private Charity charityObj;
String charitiesName[] = ["Feeding America", "Red Cross", etc];
for (String charityName: charitiesName ) {
charityObj = db.getCharityByName(charityName);
charityObj.setTrendingScoreCharity(0);
}

Isn't this what you are trying to do? Why do you need to initialize the obj;

Reply:
You can re-use one charity object for the whole loop, rite? Until and unless i am missing that we are using it later in code.

Reply:
whoops i see you are using it, later in the test.

Reply:
ok, then leave it. Although you can initialize all objects to NULL but anyways it wont look cleaner.

Reply:
I use the charity objects (FA, SJ, etc) throughout the tests (specifically for creating the expected results).